thunderbird/processes/wps_update_metadata.py

```diff
@@ -92,6 +95,37 @@ def copy_and_get_filepath(self, request):
         shutil.copyfile(original, copy)
         return copy
 
+    def updates_instruction_generator(self, request):
+        """
+        This function generates a dictionary containing the updates instruction information.
+        The dictionary will be constructed using the content of the input updates_file/updates_string.
+        It is desired that only one of the options is provided by the user, but if both are given, 
+        updates_file input will be used to generate whiile updates_string input will be neglected.
+
+        There are odd behaviours of the ComplexInput updates_file; the content of the yaml file is 
+        directly accessible using the .data attribute when run in localhost while it is accessible 
+        as '.url' attribute when run in docker containers. When running pytest, the .data attribute
+        holds the path to the origial file path.
+        """
+        if "updates_file" in request.inputs.keys():
+            try:
+                updates = request.inputs["updates_file"][
+                    0
+                ].data  # For running in localhost
+            except exceptions.NoApplicableCode:
+                updates = request.inputs["updates_file"][
+                    0
+                ].url  # For running in Docker containers
```

Comment:
While this block works fine, I believe it could be improved. We may not want to use `try/except` in this manner as we are essentially saying "if this doesn't work, try this and if that doesn't work, you're out of lucky". So rather than trying something we expect won't work, we could run a check before trying to extract the data. Perhaps we can see if `data` exists/is empty before trying to use it, or vice versa with `url`.

Reply:
I already tried if-else statement on this but wouldn't work. Checking the existence of `.data` attribute causes the `pywps.exceptions.NoApplicableCode`.

Reply:
Is there a way we could do something like:
```
if data in inputs[""].__dir__()
```
Or does that also cause an error?

Reply:
I will give that a try.

Reply:
The problem is that the attribute `data`  of the object always exists, but accessing it with `request.inputs["update_file"][0] causes an error.
```
thunderbird_test | 2020-10-09 17:11:06 CRITICAL: thunderbird: ['identifier', 'title', 'abstract', 'keywords', 'min_occurs', 'max_occurs', 'metadata', 'translations', '_workdir', '_file', '_data', '_post_data', '_stream', '_url', 'data_set', 'valid_mode', 'as_reference', 'inpt', 'uuid', '_data_format', '_supported_formats', '_default', '_default_type', 'method', '__module__', '__doc__', 'prop', 'url', 'file', 'post_data', '_openurl', 'max_input_size', 'data', 'base64', 'stream', 'mem', '_openmode', '__init__', '_reset_cache', '_check_valid', 'workdir', 'validator', 'source_type', '_set_default_value', '_build_file_name', 'extension', 'clone', '_create_fset_properties', '__dict__', '__weakref__', '__repr__', '__hash__', '__str__', '__getattribute__', '__setattr__', '__delattr__', '__lt__', '__le__', '__eq__', '__ne__', '__gt__', '__ge__', '__new__', '__reduce_ex__', '__reduce__', '__subclasshook__', '__init_subclass__', '__format__', '__sizeof__', '__dir__', '__class__', 'json', 'from_json', '_json_data', '__slotnames__', 'file_handler', 'url_handler', 'process', '_validate_file_input', 'get_format', 'supported_formats', 'data_format', '_is_supported']
thunderbird_test | 2020-10-09 17:11:06 ERROR: thunderbird: Exception: code: 400, description: File reference error: No connection adapters were found for 'global:\n    - institute_id:\n    - institute_ID: PCIC\n    - address: University House 1 \n    - Institution: <- institution\n', locator: 
thunderbird_test | Traceback (most recent call last):
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/pywps/inout/basic.py", line 450, in file
thunderbird_test |     reference_file = self._openurl(self.url, self.post_data)
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/pywps/inout/basic.py", line 491, in _openurl
thunderbird_test |     req = requests.get(url=href, stream=True)
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/requests/api.py", line 76, in get
thunderbird_test |     return request('get', url, params=params, **kwargs)
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/requests/api.py", line 61, in request
thunderbird_test |     return session.request(method=method, url=url, **kwargs)
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 530, in request
thunderbird_test |     resp = self.send(prep, **send_kwargs)
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 637, in send
thunderbird_test |     adapter = self.get_adapter(url=request.url)
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 730, in get_adapter
thunderbird_test |     raise InvalidSchema("No connection adapters were found for {!r}".format(url))
thunderbird_test | requests.exceptions.InvalidSchema: No connection adapters were found for 'global:\n    - institute_id:\n    - institute_ID: PCIC\n    - address: University House 1 \n    - Institution: <- institution\n'
thunderbird_test | Traceback (most recent call last):
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/pywps/inout/basic.py", line 450, in file
thunderbird_test |     reference_file = self._openurl(self.url, self.post_data)
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/pywps/inout/basic.py", line 491, in _openurl
thunderbird_test |     req = requests.get(url=href, stream=True)
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/requests/api.py", line 76, in get
thunderbird_test |     return request('get', url, params=params, **kwargs)
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/requests/api.py", line 61, in request
thunderbird_test |     return session.request(method=method, url=url, **kwargs)
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 530, in request
thunderbird_test |     resp = self.send(prep, **send_kwargs)
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 637, in send
thunderbird_test |     adapter = self.get_adapter(url=request.url)
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 730, in get_adapter
thunderbird_test |     raise InvalidSchema("No connection adapters were found for {!r}".format(url))
thunderbird_test | requests.exceptions.InvalidSchema: No connection adapters were found for 'global:\n    - institute_id:\n    - institute_ID: PCIC\n    - address: University House 1 \n    - Institution: <- institution\n'
thunderbird_test | 
thunderbird_test | During handling of the above exception, another exception occurred:
thunderbird_test | 
thunderbird_test | Traceback (most recent call last):
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/pywps/app/Process.py", line 249, in _run_process
thunderbird_test |     self.handler(wps_request, wps_response)  # the user must update the wps_response.
thunderbird_test |   File "/code/thunderbird/processes/wps_update_metadata.py", line 142, in _handler
thunderbird_test |     updates_instruction = self.updates_instruction_generator(request)
thunderbird_test |   File "/code/thunderbird/processes/wps_update_metadata.py", line 115, in updates_instruction_generator
thunderbird_test |     0
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/pywps/inout/basic.py", line 304, in data
thunderbird_test |     with open(self.file, mode=openmode, **kwargs) as fh:
thunderbird_test |   File "/usr/local/lib/python3.7/site-packages/pywps/inout/basic.py", line 453, in file
thunderbird_test |     raise NoApplicableCode('File reference error: {}'.format(e))
thunderbird_test | pywps.exceptions.NoApplicableCode: 400 NoApplicableCode: File reference error: No connection adapters were found for 'global:\n    - institute_id:\n    - institute_ID: PCIC\n    - address: University House 1 \n    - Institution: <- institution\n'
thunderbird_test | 2020-10-09 17:11:06 ERROR: thunderbird: Process error: method=wps_update_metadata.py._handler, line=142, msg=400 NoApplicableCode: File reference error: No connection adapters were found for 'global:\n    - institute_id:\n    - institute_ID: PCIC\n    - address: University House 1 \n    - Institution: <- institution\n'
```

Reply:
Ah ok that is interesting. I thought if there was no value it would not be there at all. I guess we could use `vars()` and check whether `_data: None`? Or does that still cause an issue?

Reply:
The output of `vars()` is in the comment above, but I will attach it here as well: 

`thunderbird_test | 2020-10-09 17:22:18 CRITICAL: thunderbird: {'identifier': 'updates_file', 'title': 'Updates File(yaml)', 'abstract': 'The filepath of an updates file that specifies what to do to the metadata it finds in the NetCDF file', 'keywords': [], 'min_occurs': 0, 'max_occurs': 1, 'metadata': [], 'translations': None, '_workdir': '/tmp/pywps_process_9ir7r7o6', '_file': None, '_data': None, '_post_data': None, '_stream': None, '_url': 'global:\n    - institute_id:\n    - institute_ID: PCIC\n    - address: University House 1 \n    - Institution: <- institution\n', 'data_set': True, 'valid_mode': 0, 'as_reference': False, 'inpt': {'identifier': 'updates_file', 'updates_file': None, 'href': 'global:\n    - institute_id:\n    - institute_ID: PCIC\n    - address: University House 1 \n    - Institution: <- institution\n', 'mimeType': 'text/x-yaml', 'method': 'GET'}, 'uuid': None, '_data_format': <pywps.inout.formats.Format object at 0x7fe46c31d390>, '_supported_formats': (<pywps.inout.formats.Format object at 0x7fe46c31d390>,), '_default': None, '_default_type': 3, 'method': 'GET'}`